waiter/src/waiter/token.clj

```diff
@@ -589,26 +623,74 @@
    If handling POST, validates that the user is the creator of the token if it already exists.
    Then, updates the configuration for the token in the database using the newest password."
   [clock synchronize-fn kv-store cluster-calculator token-root history-length limit-per-owner waiter-hostnames entitlement-manager
-   make-peer-requests-fn validate-service-description-fn attach-service-defaults-fn {:keys [request-method] :as request}]
+   make-peer-requests-fn validate-service-description-fn attach-service-defaults-fn tokens-update-chan {:keys [request-method] :as request}]
   (try
     (case request-method
       :delete (handle-delete-token-request clock synchronize-fn kv-store history-length waiter-hostnames entitlement-manager
-                                           make-peer-requests-fn request)
+                                           make-peer-requests-fn tokens-update-chan request)
       :get (handle-get-token-request kv-store cluster-calculator token-root waiter-hostnames request)
       :post (handle-post-token-request clock synchronize-fn kv-store cluster-calculator token-root history-length limit-per-owner
                                        waiter-hostnames entitlement-manager make-peer-requests-fn validate-service-description-fn
-                                       attach-service-defaults-fn request)
+                                       attach-service-defaults-fn tokens-update-chan request)
       (throw (ex-info "Invalid request method" {:log-level :info :request-method request-method :status http-405-method-not-allowed})))
     (catch Exception ex
       (utils/exception->response ex request))))
 
+(defn- handle-list-tokens-watch
+  [index-filter-fn metadata-transducer-fn tokens-watch-channels-update-chan {:keys [ctrl]}]
+  (let [watch-chan-xform
+        (comp
+          (map
+            (fn event-filter [{:keys [object type] :as event}]
+              (log/info "received event from token-watch-maintainer daemon" {:event event})
+              (case type
+                :INITIAL
+                (assoc event :object (->> object
+                                          (filter index-filter-fn)
+                                          (map metadata-transducer-fn)))
+                :EVENTS
+                (assoc event :object (->> object
+                                          (filter (fn [{:keys [object]}]
+                                                    (index-filter-fn object)))
+                                          (map (fn [{:keys [type] :as entry}]
+                                                 (if (= type :UPDATE)
+                                                   (update entry :object metadata-transducer-fn)
+                                                   entry)))))
+                (throw (ex-info "Invalid event type provided" {:event event})))))
```

Comment:
Does throwing the exception perform the necessary cleanup? Are we relying on `ctrl` chan to be triggered when the exception is thrown?

Reply:
These exceptions trigger the `watch-chan-ex-handler-fn` https://clojuredocs.org/clojure.core.async/chan